### PR TITLE
%TypedArray%.prototype.reduce() and reduceRight()

### PR DESCRIPTION
**Pull request by jtenner**
Since the `./std/assembly/internal/typedarray.ts` source changed, I took it as an opportunity to implement this function instead because the implementation was way easier to check.

Still a problem, however is the `TypedArray<u8, u32>#.__set already exists` problem.

```
Testing compiler/std/typedarray.ts

Fatal: Module::addFunction: $~lib/internal/typedarray/TypedArray<u8,u32>#__set already exists
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! assemblyscript@0.5.0 test:compiler: `node tests/compiler "--create"`
npm ERR! Exit status 1
```

This pull request is way better documented and I tried to adhere to formatting standards.

Thank you for your time and awesome work.

### CODE REVIEW

**Comment by dcodeIO** [comment]
> Still a problem, however is the TypedArray<u8, u32>#.__set already exists problem.

This should be fixed now. Let me know if it isn't.

**Comment by jtenner** [comment]
@dcodeIO This branch contains the changes you made. I think this patch is ready to submit.

File: std/assembly/internal/typedarray.ts [hunk 1]
```diff
@@ -126,4 +126,29 @@ export abstract class TypedArray<T,TNative> {
...
+   */
+  @inline
+  protected reduce_internal<TypedArrayType, ReturnType>(
+    callbackfn: (accumulator: ReturnType, value: T, index: i32, array: TypedArrayType) => ReturnType,
+    array: TypedArrayType,
+    initialValue: ReturnType,
+    ): ReturnType {
+    var index: i32 = 0;
+    var length: i32 = this.length;
+    while (index < length) {
```
Comment (dcodeIO) [inline std/assembly/internal/typedarray.ts]: What happens here if the user-provided callback somehow decides to modify the array's length? Should the loop maybe use `this.length` directly to handle this situation?

Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: TypedArrays according to the specification should not change length, ever. Am I missing something?

Comment (dcodeIO) [reply std/assembly/internal/typedarray.ts]: Oh, yeah, you are right. Just one of those reflexes left over from normal arrays on my side :)

File: std/assembly/typedarray.ts [hunk 2]
```diff
@@ -34,6 +80,29 @@ export class Uint8ClampedArray extends TypedArray<u8,u32> {
...
     return changetype<Uint8ClampedArray>(super.subarray(begin, end));
   }
+
+  /**
+   * @param callbackfn {function} - a function that reduces each value to a ReturnType
+   * @param initialvalue {ReturnType} - the initial ReturnType value to be passed to the callbackfn
+   */
+  reduce<ReturnType>(
+    callbackfn: (accumulator: ReturnType, value: u8, index: i32, array: Uint8ClampedArray) => ReturnType,
+    initialvalue: ReturnType,
```
Comment (MaxGraey) [inline std/assembly/typedarray.ts]: plz change to `initialValue` for consistency

Comment (jtenner) [reply std/assembly/typedarray.ts]: Actually I copied some stuff from mdn directly and it was a formatting issue. Thanks for the catch.

File: std/assembly/typedarray.ts [hunk 3]
```diff
@@ -98,4 +327,27 @@ export class Float64Array extends TypedArray<f64,f64> {
...
   }
+
+  /**
+   * @param callbackfn {function} - a function that reduces each value to a ReturnType
+   * @param initialValue {ReturnType} - the initial ReturnType value to be passed to the callbackfn
+   */
+  reduce<ReturnType>(
+    callbackfn: (accumulator: ReturnType, value: f64, index: i32, array: Float64Array) => ReturnType,
+    initialValue: ReturnType,
+    ): ReturnType {
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: Just a minor one, but compared to other code, this should indent by 2 less.

File: std/assembly/typedarray.ts [hunk 4]
```diff
@@ -98,4 +327,27 @@ export class Float64Array extends TypedArray<f64,f64> {
...
+    return super.reduce_internal<Float64Array, ReturnType>(callbackfn, this, initialValue);
+  }
+
+  /**
+   * @param callbackfn {function} - a function that reduces each value to a ReturnType from the end
+   * of the array instead of from the beginning
+   * @param initialValue {ReturnType} - the initial ReturnType value to be passed to the callbackfn
+   */
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: I think the comments here don't bring much to the table. The documentation a user sees for all the standard library implementaions is that in `std/assembly/index.d.ts` anyway and someone hacking on the source will most likely know what these functions are doing. Also note that the standard library sources are bundled with asc, and add to its size.

**Comment by dcodeIO** [comment]
Thanks, looking good to me with the final minor comments above addressed :)

File: std/assembly/typedarray.ts (hunk 3 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: I'm sorry, I don't understand?

File: std/assembly/typedarray.ts (hunk 4 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: Okay. I can remove the comments here.

File: std/assembly/typedarray.ts (hunk 3 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: I can format this code anyway that you would like. Can I just have a short example?

Comment (jtenner) [reply std/assembly/typedarray.ts]: Just committed. @dcodeIO please review.

Thanks!

File: std/assembly/typedarray.ts (hunk 4 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: This should be completed.

**Comment by dcodeIO** [comment]
Just noticed that there are a couple changes to package-lock that I don't know exactly where these come from, possibly a varying platform. Can we keep the original?

**Comment by jtenner** [comment]
:tada:

**Comment by dcodeIO** [comment]
Thank you! :)

**Pull request merged**